blog: Legacy projects with unsupported PHP/MySQL versions using DDEV (#375)
ddev/ddev.com: src/content/blog/legacy-projects-with-unsupported-php-and-mysql-using-ddev.md

---
title: "Using DDEV to spin up a legacy PHP application"

> pubDate: 2025-05-27
> modifiedDate: 2025-05-27

Reviewer: ```suggestion
pubDate: 2025-05-29
#modifiedDate: 2025-05-29
```

One advantage (of many) of using DDEV instead the underlying Docker Compose is that so many things are pre-configured and "just work". So I really did not want to migrate everything to Docker Compose on my own, do my custom routing, PHP-FPM integration and whatnot.

Just being able to "bait and switch" the PHP and DB container with a different base Docker image was all that was needed for me:

## Step 1: Base config

I created the base `~/legacyphp/.ddev/config.yaml` file manually inside my `~/legacyphp` project directory, setting `legacyphp` as the project name.

Note that I configured PHP and MySQL versions that are supported by DDEV for this first:

```yaml
name: legacyphp
type: php
docroot: htdocs
php_version: "8.3"
webserver_type: apache-fpm
database:
  type: mysql
  version: "8.0"
```

## Step 2: Rewire DB

Next I created the very small file `~/legacyphp/.ddev/docker-compose.db.yaml` in the same directory next to `config.yaml`:

```yaml
services:
  db:
    platform: linux/amd64
    build:
      args:
        BASE_IMAGE: ddev/ddev-dbserver-mysql-5.5:v1.24.6

> ```

Reviewer: Trick DDEV into thinking it's MySQL 8.0:

````suggestion
    entrypoint: ["sh", "-c", "cp /docker-entrypoint.sh ~/docker-entrypoint.sh; sed -i '157s/.*/if false; then/' ~/docker-entrypoint.sh; sed -i '175s|.*|echo 'mysql_8.0' >/var/lib/mysql/db_mariadb_version.txt|' ~/docker-entrypoint.sh; exec ~/docker-entrypoint.sh"]
```
````

Reviewer: This is the file that was edited https://github.com/ddev/ddev/blob/main/containers/ddev-dbserver/files/docker-entrypoint.sh

---
title: "Using DDEV to spin up a legacy PHP application"
pubDate: 2025-05-27
modifiedDate: 2025-05-27
summary: How to use DDEV with outdated PHP and MySQL versions
author: Garvin Hicking
featureImage:
  src: /img/blog/2025/05/museums-victoria-Di7WfLcrJ_I-unsplash.jpg
  alt: Legacy computer museum

>   credit: "Photo by [Museums Victoria](https://unsplash.com/@museumsvictoria?utm_content=creditCopyText&utm_medium=referral&utm_source=unsplash) on [Unsplash](https://unsplash.com/photos/gray-mechanical-machine-lot-beside-wall-Di7WfLcrJ_I?utm_content=creditCopyText&utm_medium=referral&utm_source=unsplash)"

Reviewer: I think stuff like `?utm_content=creditCopyText&utm_medium=referral&utm_source=unsplash` is not needed.

categories:
  - Guides
---

_This guest post is by DDEV community member and [TYPO3](https://typo3.org) contributor [Garvin Hicking](/blog/author/garvin-hicking/)._

> In my daily work I am developing [TYPO3](https://typo3.org)-based projects and also contribute to the TYPO3 CMS OpenSource project itself.

Reviewer: ```suggestion
In my daily work, I develop [TYPO3](https://typo3.org)-based projects and also contribute to the TYPO3 CMS OpenSource project itself.
```

Usually this means working with actively supported and up-to-date PHP versions as well as database systems like MySQL/PostgreSQL/MariaDB.

Just recently I had to migrate a very outdated project: TYPO3 4.5, which utilized MySQL 5.5 and PHP 5.3. When that project was initially developed, it was done with XAMPP and later Vagrant-based VMs. This has been long superseded with using Docker and specifically DDEV for ease-of-use.

So naturally I wanted to be able to use DDEV for the legacy project to get it working just as it is running on the (outdated) hosting provider's shared web servers.

I quickly faced three major issues:

- No PHP 5.3 out-of-the-box support from DDEV; it starts with 5.6 as of the time of this writing

> - No MySQL 5.5 support either; it starts with 5.7

Reviewer: ```suggestion
- No MySQL 5.5 ARM64 support either; it starts with 5.7
```

- Additionally, I use an Apple MacBook Pro M1 with ARM-chipset, which has no "official" MySQL 5.5 support

Thanks to the outstanding DDEV support on Discord, I was quickly able to find a way with minimal effort, just by creating very small custom, additional docker-compose YAML files.

One advantage (of many) of using DDEV instead the underlying Docker Compose is that so many things are pre-configured and "just work". So I really did not want to migrate everything to Docker Compose on my own, do my custom routing, PHP-FPM integration and whatnot.

Just being able to "bait and switch" the PHP and DB container with a different base Docker image was all that was needed for me:

## Step 1: Base config

I created the base `~/legacyphp/.ddev/config.yaml` file manually inside my `~/legacyphp` project directory, setting `legacyphp` as the project name.

Note that I configured PHP and MySQL versions that are supported by DDEV for this first:

> ```

Reviewer: Please add language identifiers, such as:

````suggestion
```yaml
````

name: legacyphp
type: php
docroot: htdocs
php_version: "8.3"
webserver_type: apache-fpm
database:
    type: mysql
    version: "8.0"
```

## Step 2: Rewire DB

Next I created the very small file `~/legacyphp/.ddev/docker-compose.db.yaml` in the same directory next to `config.yaml`:

> ```

Reviewer: ````suggestion
```yaml
````

services:
  db:
    platform: linux/amd64
    build:
      args:
        BASE_IMAGE: ddev/ddev-dbserver-mysql-5.5:v1.24.6

> ```

Reviewer: This is an interesting workaround you discovered, but it's going to work only until the next `ddev restart`:

```
Failed to start legacyphp: unable to start project legacyphp because the configured database type does not match the current actual database. Please change your database type back to mysql:5.5 and start again, export, delete, and then change configuration and start. To get back to existing type use 'ddev config --database=mysql:5.5' and then you might want to try 'ddev debug migrate-database mysql:8.0', see docs at https://ddev.readthedocs.io/en/stable/users/extend/database-types/
```

I think we can probably insert `platform:linux/amd64` for MySQL 5.5, 5.6 on arm64.

I opened this issue:

- https://github.com/ddev/ddev/issues/7338

But I don't think we're going to implement this fast.

I propose to add a small note on this matter.

Author: Oooh, shoot. That is unfortunate - I honestly only had to start it up once and left it open for some days, and never issued a restart. I wonder if this makes the whole point of the article moot now, if we do not have a way to re-execute the container.

So I'd possibly scrap the whole article now, if you don't see any other kind of workaround to bypass the database config.yaml assignment/validation. If we had an YAML option like "disable_validation" or so that would massively help of course...

Author: Hm... I'm a bit afraid this "kills" the lightweightness the solution had before... I'm not so sure this would be a solution we should highlight as a blog entry then?

I would be okay to leave this blog entry dangling, if a ddev release allowing to influence the behaviour could be done at some point. Then we could add an info about the version requirement to the entry.. I am not in a hurry to publish this, but I also don't want to blow up the complexity too much.

Reviewer: Let’s put this PR in draft then.

(No new releases this month or next while Randy is bikepacking.)

Two things are noteworthy:

- Setting `linux/amd64` as the platform will require Rosetta to be available on the macOS ARM64 platform
- The `BASE_IMAGE` is set to a DDEV DB container of legacy ddev versions which are still available.

## Step 3: Rewire PHP

Using a different PHP work is just a few lines more work, because we are not replacing the whole `web` container oh DDEV, but instead add an additional PHP container which is executed from the web container via port 9000.

This is done via the file `~/legacyphp/.ddev/docker-compose.php.yaml`:

> ```

Reviewer: ````suggestion
```yaml
````

services:
  php:
    container_name: ddev-${DDEV_SITENAME}-php
    image: devilbox/php-fpm:5.3-work

    restart: "no"
    ports:
      - 9000
    labels:
      com.ddev.site-name: ${DDEV_SITENAME}
      com.ddev.approot: ${DDEV_APPROOT}
    volumes:
      - type: bind
        source: ../
        target: /var/www/html
        consistency: cached
      - ".:/mnt/ddev_config:ro"
      - ddev-global-cache:/mnt/ddev-global-cache
      - "./php:/etc/php-custom.d"
    environment:
      - DDEV_PHP_VERSION
      - IS_DDEV_PROJECT=true
  web:
    links:
    - php:php
    healthcheck:
      test: ["CMD", "true"]
```

Note here that we use `devilbox/php-fpm` with our needed version, and a bind-mount takes care the PHP container can access our main project root directory.

A special mount of `~/legacyphp/.ddev/php/` is included so that we can control the `php.ini` configuration, if needed. For example you could disable the OPCache+APC in case you're doing some legacy benchmarking that should not be falsified via caching, I created a very small file `~/legacyphp/.ddev/php/php.ini` file with the contents:

> ```

Reviewer: ````suggestion
```ini
````

# This is an example.
# apc.enabled=Off
# opcache.enable=Off
```

## Step 4: Utilize the PHP container with an Apache proxy

To execute PHP with our external PHP docker image, I created the following file in `~/legacyphp/.ddev/apache/apache-site.conf`:

> ```

Reviewer: Not sure if there is something for apache.

Author: Ah. According to https://shiki.style/languages#bundled-languages `apache` should do the trick, will try.

Author: looks good to me, it has styling using that markup

<VirtualHost *:80>
    RewriteEngine On
    RewriteCond %{HTTP:X-Forwarded-Proto} =https
    RewriteCond %{DOCUMENT_ROOT}%{REQUEST_FILENAME} -d
    RewriteRule ^(.+[^/])$ https://%{HTTP_HOST}$1/ [redirect,last]
    SetEnvIf X-Forwarded-Proto "https" HTTPS=on

    DocumentRoot /var/www/html/htdocs
    <Directory "/var/www/html/htdocs">
      AllowOverride All
      Allow from All
    </Directory>

    CustomLog /var/log/apache2/access.log combined
    ProxyFCGIBackendType GENERIC
    ProxyPassMatch ^/(.*\.php(/.*)?)$ fcgi://php:9000/var/www/html/htdocs/$1
    DirectoryIndex /index.php index.php
</VirtualHost>
```

Note that if your document root is not `htdocs` you would need to adapt this name to your liking (like `public` or `wwwroot` or anything) in all occurrences of this file.

## Step 5: Lift-Off

Now you can execute `ddev start` and then `ddev launch` to see your project up and running.

You could create a simple `~/legacyphp/htdocs/index.php` file with `<?php phpinfo(); ?>` to verify the version.

Using `ddev mysql` will connect you to the MySQL 5.5. instance:

> ```

Reviewer: ````suggestion
```bash
````
